Refactor `runPeriodicTask` helpers

masnax opened the pull request:
Primarily fixes the issue with failing to process updates to an instance from the source leading to instance record removal. But I noticed these helpers were ignoring a lot of errors, so I tried to clean them up a bit. 

For the most part I've changed these helpers so all (really, most) errors are handled, and it's up to the caller to decide to skip them. But since these are run in `runPeriodicTask`, the final error is always just logged. 

The primary change with this approach is that an error that occurs within a database transaction will now cause the transaction to be rolled back, whereas before most errors were not returned in transactions so they would be committed in all cases.

I've also split up a lot of the transactions so that the callout to the source & target are not nested in the transactions themselves. I've used a lot of maps to keep track of records fetched from the database. This also results in some efficiency improvements as we can collect all the batches, instances, and targets we care about upfront and not reach out so frequently to the database for each instance in each batch. There's still more efficiency improvements to make here but I suppose that all boils down to what the behaviour will be if we have like 10,000 instances in a batch.

I'm not sure if there's a good reason to ignore all these errors, and especially so within transactions, so if there's something big I'm missing then please do enlighten me :)

Other changes:
* Use `revert` from Incus so that we don't duplicate so many calls to set batch/instance status to errored if we see an error.
* Clean up instances that were created as part of a failed migration.

masnax commented:
@gibmat I'm leaving this as draft because the changes are pretty sweeping so I'd like to do some live testing before making it mergeable, but I would appreciate a look-over in case I'm totally on the wrong track here :)

breml commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -45,251 +45,355 @@ func (d *Daemon) runPeriodicTask(f func() bool, interval time.Duration) {
+			return fmt.Errorf("Failed to get internal network records: %w", err)
+		}
+
+		dbInstances, err = d.instance.GetAll(ctx)
+		if err != nil {
+			return fmt.Errorf("Failed to get internal instance records: %w", err)
+		}
+
+		// Build maps to make comparison easier.
+		existingNetworks := make(map[string]migration.Network, len(dbNetworks))
+		for _, net := range dbNetworks {
+			existingNetworks[net.Name] = net
+		}
+
+		existingInstances := make(map[uuid.UUID]migration.Instance, len(dbInstances))
+		for _, inst := range dbInstances {
+			existingInstances[inst.UUID] = inst
+		}
+
+		for srcName, srcInstances := range instancesBySrc {
+			srcNetworks := networksBySrc[srcName]
+			err = syncNetworksFromSource(ctx, srcName, d.network, existingNetworks, srcNetworks)
 			if err != nil {
-				log.Warn("Failed to connect to source", logger.Err(err))
-				continue
+				return err
 			}
 
-			networks, err := s.GetAllNetworks(d.ShutdownCtx)
+			err = syncInstancesFromSource(ctx, srcName, d.instance, existingInstances, srcInstances)
 			if err != nil {
-				log.Warn("Failed to get networks", logger.Err(err))
-				continue
+				return err
 			}
+		}
```
This looks pretty similar to `syncSourceData`, no? Can these two code sections be combined?

masnax replied:
Added a shared helper. Sadly to maintain the same behaviour, the signature of the shared helper is pretty long (two maps of maps).

breml commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -481,7 +598,13 @@ func (d *Daemon) ensureISOImagesExistInStoragePool(ctx context.Context, instance
 
 	// Get the target.
-	t, err := d.target.GetByID(ctx, *inst.TargetID)
+	var t migration.Target
+	err = transaction.Do(ctx, func(ctx context.Context) error {
+		var err error
+		t, err = d.target.GetByID(ctx, *inst.TargetID)
+
+		return err
+	})
```
Single statement, no need to wrap it in a transaction.

masnax replied:
This is a good example of what I'm talking about with having a bias for transactions leading to thinking about structuring code better.

You caught that this looked weird. Now as I'm looking at this block more closely, I think it seems wrong to even have db access in this helper. After all, it's just about checking if an Incus target has the right storage volumes. If anything, we should have already instantiated a target client, and this function should expect it as an argument.

breml commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -539,40 +662,54 @@ func (d *Daemon) ensureISOImagesExistInStoragePool(ctx context.Context, instance
-		log.Warn("Failed to get source by ID", logger.Err(err))
-		_, err = d.instance.UpdateStatusByUUID(ctx, inst.UUID, api.MIGRATIONSTATUS_ERROR, err.Error(), true)
+		err := transaction.Do(ctx, func(ctx context.Context) error {
+			_, err := d.instance.UpdateStatusByUUID(ctx, inst.UUID, api.MIGRATIONSTATUS_ERROR, errString, true)
+			return err
+		})
 		if err != nil {
-			log.Warn("Failed to update instance status", slog.Any("status", api.MIGRATIONSTATUS_ERROR), logger.Err(err))
+			log.Error("Failed to update instance status", slog.Any("status", api.MIGRATIONSTATUS_ERROR.String()), logger.Err(err))
 		}
+	})
```
Single statement, no need to wrap it in a transaction.

masnax replied:
This is a best-effort attempt in the reverter, so I think there's no issue with removing the transactions.

breml commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -592,120 +729,142 @@ func (d *Daemon) spinUpMigrationEnv(ctx context.Context, inst migration.Instance
-	// TODO: context should be passed from the daemon to all the workers.
-	ctx := context.TODO()
+		for _, i := range instances {
+			err := transaction.Do(ctx, func(ctx context.Context) error {
+				_, err := d.instance.UpdateStatusByUUID(ctx, i.UUID, api.MIGRATIONSTATUS_ERROR, errString, true)
+				return err
+			})
```
Shouldn't the transaction be outside of the loop or no transaction at all?

breml reviewed (commented):
What are your thoughts about the inconsistent states, that might occur with the split up transactions? How do we notice them and how we can recover from them?

gibmat commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -806,7 +826,7 @@ func (d *Daemon) finalizeCompleteInstances() bool {
 			if err != nil {
 				log.Warn("Failed to get target", slog.Int("target_id", *i.TargetID), logger.Err(err))
-				_, err := d.instance.UpdateStatusByUUID(ctx, i.UUID, api.MIGRATIONSTATUS_ERROR, err.Error(), true)
+				_, err := d.instance.UpdateStatusByID(ctx, i.UUID, api.MIGRATIONSTATUS_ERROR, err.Error(), true)
```
We don't want to undo the method renaming. :wink:

masnax replied:
Oops, bad rebase between commits :)

gibmat commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -767,6 +767,13 @@ func (d *Daemon) spinUpMigrationEnv(ctx context.Context, inst migration.Instance
 	}
 
+	reverter.Add(func() {
+		err := it.DeleteVM(instanceDef.Name)
```
There probably should be some configuration knob to enable/disable this. When a migration fails it's often necessary to exec into the failed environment to look at the worker's log and/or further poke at whatever caused the failure.

masnax commented on cmd/migration-managerd/internal/api/daemon.go:
```diff
@@ -262,10 +262,42 @@ func (d *Daemon) Start() error {
+				slog.Error("Failed to sync all sources", logger.Err(err))
+			}
+
+			return done
```
What sort of errors do you expect will suggest that we should exit the periodic task?

It seems to me these tasks should always be running and trying to perform their actions as long as the daemon is active. The clearest "done" case is if `d.ShutdownCtx` has been cancelled. But checking that should likely be handled in the implementation of `runPeriodicTask` itself, rather than the individual helpers.

breml replied:
You are right, I did not check the implementation of `runPeriodicTask` so I missed, that checking if the context is done, is handled there. This makes sense.

masnax commented on cmd/migration-managerd/internal/api/workers.go:
```diff
@@ -45,251 +45,355 @@ func (d *Daemon) runPeriodicTask(f func() bool, interval time.Duration) {
 		// Get the list of configured sources.
-		sources, err := d.source.GetAll(ctx)
+		var err error
+		sources, err = d.source.GetAll(ctx)
 		if err != nil {
 			return fmt.Errorf("Failed to get all sources: %w", err)
 		}
 
-		// Check each source for any net networks and any new, changed, or deleted instances.
-		for _, src := range sources {
-			log := log.With(slog.String("source", src.Name))
+		return nil
+	})
+	if err != nil {
+		return err
+	}
```
You're right that it's a inefficient to do it this way, but to give some context of my thought process, I find it's useful for maintainability to keep transactions in mind whenever you need database access. 

Using this function as an example, there's 3 distinct components: 
* transaction to fetch data
* make API calls using the data
* another transaction to commit new data

If or when this block of coded is extended, its existing structure makes it easier to reason about how to structure the implementation without causing fragmentation with additional db queries here and there, and then down the line having to reorganize the logical flow.

That said I'm perfectly happy to simplify them out if you think it's noisy.

breml replied:
OK, I can follow your reasoning and I can see the advantages of this approach. I guess, what triggered me was, that initially in migration manager every DB query was wrapped in its own transaction due to the fact, that the DB layer took `*db.Tx` as an argument. This way, the transactions have been basically useless, since the queries that need to go together in a transaction also have been in separate ones.

One way or the other, people working on this code need to understand, which actions need to go in a transaction in order to maintain consistency and which operations are save to be run on their own (e.g. without a transaction or in a transaction with just a single query).

masnax replied:
I've left this one in place since it fits in smoothly, but in other places I've avoided single-query transactions. This should be the only one left.

masnax commented:
> What are your thoughts about the inconsistent states, that might occur with the split up transactions? How do we notice them and how we can recover from them?

On the one hand, previously errors were being ignored in these transactions so they offered very little guarantee of consistency. Now we have proper rollback within transactions so that's one area of improved consistency. 

With the transactions split up, we do introduce some new problems that we should consider:
* If we commit a transaction and encounter an error later, we have to manually roll back the changes so that the whole task can be retried. We have the `reverter` that can help with this. The pattern of reading data upfront and only writing at the end of the function also helps with this, because all the writes are usually grouped together and thus will rollback together on errors.

* We have less control over the operation order because another process might start a transaction and read/write data in between a running task's transactions. This isn't too much of a concern because we're already quite strict with how the API can be used to update resources, but looking at each helper individually:
  * `trySyncAllSources` is safe because the first transaction just fetches sources, and the remainder of the function operates on only those sources. If a source is deleted by another process right after we fetched its data but before we get to the second transaction, then the sync will fail and rollback, and be tried again later.
  
  * `processReadyBatches` and and `finalizeCompleteInstances` are safe too, since the only write is setting the batch/instance state for the given set of batches/instances at the end of the functions. If a new qualifying batch/instance appears while the function is running, it will be captured by the next run in 10s. If a batch/instance that we expected to set the status of is deleted while in queued or ready state, then the functions will run the `reverter` and set the status to `ERROR`.
  
  * `processQueuedBatches` is actually a problematic one because of the spawned goroutine. Looking at this more, it can actually be broken down further. This task should end in setting the instance state to `IDLE` and then we should have a new periodic task that moves instances from `IDLE` to `FINISHED`, rather than nesting a goroutine here.

breml commented:
> * If we commit a transaction and encounter an error later, we have to manually roll back the changes so that the whole task can be retried. We have the `reverter` that can help with this. The pattern of reading data upfront and only writing at the end of the function also helps with this, because all the writes are usually grouped together and thus will rollback together on errors.

Just something to keep in mind, even with the `reverter` we might end up in an undefined or inconsistent state, since it is not guaranteed, that the `reverter` is successful in reverting the previous operation or that a certain operation even can be reverted. But for sure, it is better to try than to simply ignore. That being said, we need to make sure, that such cases can "easily" be discovered in the logs, maybe we can even "mark" them with a defined field, such that we can easily search for these cases.

> ... but looking at each helper individually:

I like your reasoning about each helper. 👍🏻 

>   * `trySyncAllSources` is safe because the first transaction just fetches sources, and the remainder of the function operates on only those sources. If a source is deleted by another process right after we fetched its data but before we get to the second transaction, then the sync will fail and rollback, and be tried again later.
>   * `processReadyBatches` and and `finalizeCompleteInstances` are safe too, since the only write is setting the batch/instance state for the given set of batches/instances at the end of the functions. If a new qualifying batch/instance appears while the function is running, it will be captured by the next run in 10s. If a batch/instance that we expected to set the status of is deleted while in queued or ready state, then the functions will run the `reverter` and set the status to `ERROR`.
>   * `processQueuedBatches` is actually a problematic one because of the spawned goroutine. Looking at this more, it can actually be broken down further. This task should end in setting the instance state to `IDLE` and then we should have a new periodic task that runs the workers on `IDLE` instances and sets their status to `IMPORT_COMPLETE`, rather than nesting a goroutine here.

I wonder, if we should add some of the thoughts you have put together as comments to the code.

masnax commented:
So for round two, I've refactored `processQueuedBatches` pretty substantially, and added some comments to describe the process. 

The main concern was the spawned goroutines at the end of the function. Those are all gone now, and instead `processQueuedBatches` just sets instance and batch states, and does so slightly more strictly.

Every 10s, in processQueuedBatches:
 * Get all batches
 * Filter only the batches we care about (QUEUED, in window)
 * Filter only the instances we care about (ASSIGNED)
 * Set batches to RUNNING, and their instances to CREATING

The migration worker setup has been split off into its own periodic task so we can track its completion, and there's a generic `RunConcurrentList` helper that spawns however many goroutines we need and waits for them to complete. 

Every 10s, in initMigrationWorkers:
 * Get all RUNNING batches and their CREATING instances
 * Concurrently:
   * Create VMs on each instance
   * Exec the migration worker for each instance
   * If an error occurs, set the instance to ERRORED
   * Optionally delete the target VM

Before, a new `processQueuedBatches` and the previous run's spawned goroutines could be running in parallel, leading to a bit of non-determinism. Now, we still trigger the instance creation in parallel, but we have a dedicated goroutine waiting for them to finish.

By using instance and batch states, we can further ensure that the runs don't interfere with each other:
* processQueuedBatches takes QUEUED batches and ASSIGNED instances and makes them RUNNING and CREATING, respectively 
* initMigrationWorkers expects precisely these states, so it will skip any batches that are modified by some other goroutine. If it fails or succeeds, the instance states will be different and so there's no risk of it running again (unless some database error prevents the state setting).

> Just something to keep in mind, even with the `reverter` we might end up in an undefined or inconsistent state, since it is not guaranteed, that the `reverter` is successful in reverting the previous operation or that a certain operation even can be reverted. But for sure, it is better to try than to simply ignore. That being said, we need to make sure, that such cases can "easily" be discovered in the logs, maybe we can even "mark" them with a defined field, such that we can easily search for these cases.

I've added some basic tagging to each log in the reverters, so we can distinguish between generic error logs and reverter error logs. Maybe we can have a wrapper around the reverter to always do this for us so we don't have to remember to.

stgraber reviewed (approved):
Pretty high level review of the various commits, but the general structure seems good and given this was tested to still work in the test environment, logic should be generally correct :)
